---
title: "07_DESeq2_Sed_D4-D66"
author: "Sara Correa Garcia"
format: html
---

# -------------------------------------------------------

# Data wrangling

# -------------------------------------------------------
Use the genus dataset raw data
 
## Load packages
```{r}
library(here)
source(here("source", "libraries.R"))
```

## Load data
```{r}
#load(here("RData", "RelAbund_tables_RData"))
load(here("RData", "Sediment_data.RData"))
raw <- read.csv(file = here("data", "clean", "feature_table_filtered.tsv"), dec = ".", sep = "\t", header = T, row.names = 1, comment.char = "") #20078 objs in 299 vars
```

## Keep only experimental samples (E2)
```{r}
# Find columns with "E2" in their name
e2_cols <- grep("E2", colnames(raw))
# Find the index of the "taxonomy" column
taxonomy_col <- which(colnames(raw) == "taxonomy")
# Combine the indices
cols_to_keep <- c(e2_cols, taxonomy_col)
# Subset the dataframe
raw <- raw[, cols_to_keep] # 20078 and 284
```


## Keep only ASVs from samples
ASV present ONLY in negative controls will be excluded at this point
```{r}
# Compute row sums excluding the 'taxonomy' column
row_sums_without_taxonomy <- rowSums(raw[, sapply(raw, is.numeric)])

# Subset the dataframe based on the condition
raw <- raw[row_sums_without_taxonomy > 0, ] # 20063 and 284
```


Quickly fix sample names
```{r}
names(raw) <- sub("X", "", names(raw))
dim(raw) # 20063   284 including taxonomy
```

Check library size
```{r}
raw_libsize <- as.data.frame(colSums(raw[,1:ncol(raw)-1]))
summary(raw_libsize)
```

There is one library that is too small: 223_224MesoRh_4_1G_E2_16S_D66 with only 10 counts_ The other samples_
 Min_   :   10                  
 1st Qu_:20850                  
 Median :38555                  
 Mean   :37614                  
 3rd Qu_:53453                  
 Max_   :99100  

Eliminate 223.224MesoRh.4.1G.E2.16S.D66 from raw and metadata

```{r}
'223.224MesoRh.4.1G.E2.16S.D66' %in% colnames(raw)
raw <- raw[, colnames(raw) != '223.224MesoRh.4.1G.E2.16S.D66']
meta_sed <- meta_sed[rownames(meta_sed) != '223.224MesoRh.4.1G.E2.16S.D66',]
```

tidy environment
```{r}
rm(list = c("cols_to_keep", "e2_cols", "row_sums_without_taxonomy", "taxonomy_col", "raw_libsize"))
```

## Keep only SEDIMENTS
```{r}
raw_sed <- raw[, colnames(raw) %in% rownames(meta_sed)] #Subset columns
#raw_sed <- cbind(raw_sed, raw$taxonomy)
raw_sed <- raw_sed[(rowSums(raw_sed) > 0) ,] # Keep only ASV with at leat 1 count_
```

Evaluate library size for samples and evaluate the effect of small library size in D4 samples
```{#r}
raw_sed_libsize <- as.data.frame(colSums(raw_sed[,1:ncol(raw_sed)]))
summary(raw_sed_libsize) # evaluate library size: samples 25, 27, 30, 31 and 8 are too small. 
meta_sed_4 <- meta_sed[meta_sed$Time == "D4",] # Subset D4 in meta
d4 <- raw_sed[,colnames(raw_sed) %in% rownames(meta_sed_4)] # subset d 4 in raw_sed 188851 obs in 24 vars
d4 <- d4[(rowSums(d4) > 0) ,] # 3735 obs in 24 vars, Keep only ASV with at least 1 count
```


#Create an ordination to see if the samples are clustering together with the other D4 samples
```{#r}
d4_t <- t(d4) # transpose
d4_s <- d4_t[order(rownames(d4_t)),] # order
rownames(d4_s) == rownames(meta_sed_4) # check names
d4_rel <- d4_s/rowSums(d4_s) # calculate relative abundance
rowSums(d4_rel)
d4_h <- decostand(d4_s, "hellinger")
d4_pca <- prcomp(d4_h, scale = TRUE)
screeplot(d4_pca, npcs = 20, type = "lines", bstick = T)
df <- iris[c(1, 2, 3, 4)]
autoplot(prcomp(d4_h), scale = TRUE, shape = FALSE, label.size = 2)
ggsave(filename = here("output", "figures", "d4_test_small_library_size_effect.eps"),
       plot = last_plot(),
       width = 7,
       height = 5)
```


extract taxonomy
```{r}
taxonomy <- as.data.frame(raw$taxonomy, row.names = rownames(raw)) # get taxonomy appart
taxonomy <- as.data.frame(taxonomy[rownames(taxonomy) %in% rownames(raw_sed),], row.names = rownames(raw_sed))
```


## Clean taxonomy

Separate tax_raw into groups
```{#r}
tax_raw_clean <- separate(taxonomy, taxonomy, into = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain"), sep = ";")
tax_raw_clean <- tax_raw_clean[,-which(names(tax_raw_clean) %in% c("Strain"))]
tax_raw_clean <- as.data.frame(tax_raw_clean)
levels(as.factor(tax_raw_clean$Domain)) # 2 Domain: Archaea and Bacteria, all ok
levels(as.factor(tax_raw_clean$Phylum)) #  43 Phyla 
levels(as.factor(tax_raw_clean$Class)) # 121 classes
levels(as.factor(tax_raw_clean$Order)) # 296 orders
levels(as.factor(tax_raw_clean$Family)) # 466 Families
levels(as.factor(tax_raw_clean$Genus)) # 882 genus
str(tax_raw_clean)
```

Clean taxonomy names
```{#r shorter version to be applied to DRY up the process below}
# eliminate the characters  -----
pattern <- "[a-z]__"
replacement <- ""
tax_clean <- as.data.frame(apply(tax_raw_clean, 2, function(col) gsub(pattern, replacement, col)))
rm(pattern)
rm(replacement)
```

Fill NAs with last available taxon rank
```{#r old version}
tax_clean_filled <- tax_clean |> 
   t() |> 
   as.data.frame() |> 
   fill(everything(), _direction = "down") |> 
   t() |> 
   as.data.frame()
```


Clean taxonomy green genes format
```{r}

colnames(taxonomy) <- c("taxonomy")
# clean the taxonomy, Greengenes format
tax <- taxonomy |> 
  select(taxonomy) |> 
  separate(taxonomy, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax_clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)
tax_clean$Species <- NULL
tax_clean[is.na(tax_clean)] <- ""
tax_clean[tax_clean == "__"] <- ""

#Upon inspection, there are several "uncultured" located at different taxonomic ranges_ 
tax_clean[tax_clean == "uncultured"] <- ""
tax_clean[tax_clean == " "] <- ""

for (i in 1:nrow(tax_clean)) {
   if (tax_clean[i,2] == "") {
    kingdom <- paste("Unclassified", tax_clean[i,1], sep = " ")
    tax_clean[i, 2:6] <- kingdom
  } else if (tax_clean[i,3] == "") {
    phylum <- paste("Unclassified", tax_clean[i,2], sep = " ")
    tax_clean[i, 3:6] <- phylum
  } else if (tax_clean[i,4] == "") {
    class <- paste("Unclassified", tax_clean[i,3], sep = " ")
    tax_clean[i, 4:6] <- class
  } else if (tax_clean[i,5] == "") {
    order <- paste("Unclassified", tax_clean[i,4], sep = " ")
    tax_clean[i, 5:6] <- order
  } else if (tax_clean[i,6] == "") {
    family <- paste("Unclassified", tax_clean[i,5], sep = " ")
    tax_clean[i, 6:6] <- family
  } 
}
```



## Sanity check for genus
```{r}
if (all(row.names(tax_clean) == rownames(raw_sed) )) {
  print("Sanity check ok")
} else {
  print("ERROR: not all row names are the same")
}
```

## Sanity check for meta_sed
```{r}
if (all(row.names(meta_sed) == colnames(raw_sed) )) {
  print("Sanity check ok")
} else {
  print("ERROR: not all row names are the same")
}
```

# -------------------------------------------------------

# DESeq2 - Phyloseq

# -------------------------------------------------------

## Create phyloseq object

```{r}
library(phyloseq)
otu <- otu_table(as.matrix(raw_sed), taxa_are_rows = TRUE)
tax <- tax_table(as.matrix(tax_clean))
sample <- sample_data(meta_sed)

# merge
ps <- phyloseq(otu, tax, sample)
sample_data(ps)$Time <- as.factor(sample_data(ps)$Time) # factorize for DESeq2
```
 ## Run only once
```{#r}
save(ps, tax_clean, file = here("RData", "ps.RData"))
```

## Create phyloseq object wit taxonomy as a single string

```{r}
library(phyloseq)
otu <- otu_table(as.matrix(raw_sed), taxa_are_rows = TRUE)
tax <- tax_table(as.matrix(taxonomy))
sample <- sample_data(meta_sed)

# merge
ps_string <- phyloseq(otu, tax, sample)
sample_data(ps_string)$Time <- as.factor(sample_data(ps_string)$Time) # factorize for DESeq2
```
Run only once
```{#r}
save(ps_string, file = here("RData", "ps_string.RData"))
```


# -------------------------------------------------------

# DESeq2 - GENUS

# -------------------------------------------------------
```{r}
load(file = here::here("RData", "ps.RData"))
```

## Agglom at genus
```{r}
ps_taxa <- tax_glom(ps, taxrank = 'Genus', NArm = FALSE) # 1046
```


## DESeq2 genus
```{r}
# Subset the original phyloseq object 'ps_taxa' to only include samples from time points "D4" and "D66"
ps_taxa_sub <- subset_samples(ps_taxa, Time %in% c("D4", "D66"))

# Filter out features (OTUs or taxa) that have more than 90% zeros across all samples
# This is done by checking the number of zeros in each row of the OTU table and comparing it to 90% of the total number of columns (samples)
ps_taxa_pse_sub <- prune_taxa(rowSums(otu_table(ps_taxa_sub) == 0) < ncol(otu_table(ps_taxa_sub)) * 0.9, ps_taxa_sub) #612

# Convert the filtered phyloseq object to a DESeq2 object, using 'Time' as the variable of interest
ps_ds <- phyloseq_to_deseq2(ps_taxa_pse_sub, ~ Time)

# Estimate size factors for normalization using the "poscounts" method, which is useful when many genes have zeros in some samples
ds <- estimateSizeFactors(ps_ds, type = "poscounts")

# Perform differential abundance analysis using the Wald test and a parametric fit
ds <- DESeq(ds, test = "Wald", fitType = "parametric")

# Set the significance threshold (alpha) to 0_05
alpha <- 0.05 

# Retrieve the results of the differential abundance analysis
res <- results(ds, alpha = alpha)

# Order the results by the adjusted p-values (padj) in ascending order
res <- res[order(res$padj, na.last = NA), ]

# Extract the names (taxa) of the top 20 features with the lowest adjusted p-values
taxa_sig <- rownames(res[1:20, ])

# Convert the counts in the original phyloseq object 'ps_taxa' to relative abundances (percentages)
ps_taxa_rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)

# Subset the relative abundance phyloseq object to only include the significant taxa identified earlier
ps_taxa_rel_sig <- prune_taxa(taxa_sig, ps_taxa_rel)

# Further subset the relative abundance phyloseq object to only include samples from time points "D4" and "D66"
ps_taxa_rel_sig <- prune_samples(colnames(otu_table(ps_taxa_pse_sub)), ps_taxa_rel_sig)

```

## Heatmap genus
```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)

matrix <- as.matrix(data.frame(otu_table(ps_taxa_rel_sig)))
rownames(matrix) <- as.character(tax_table(ps_taxa_rel_sig)[, "Genus"])
metadata_sub <- data.frame(sample_data(ps_taxa_rel_sig))
# Define the annotation color for columns and rows
annotation_col = data.frame(
    `Plant type` = as.factor(metadata_sub$Plant.type), 
    Time = as.factor(metadata_sub$Time), 
    check.names = FALSE
)
rownames(annotation_col) = rownames(metadata_sub)

annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_taxa_rel_sig)[, "Phylum"])
)
rownames(annotation_row) = rownames(matrix)

# ann_color should be named vectors
phylum_col = RColorBrewer::brewer.pal(length(levels(annotation_row$Phylum)), "Paired")
names(phylum_col) = levels(annotation_row$Phylum)
ann_colors = list(
    `Plant type` = c(`no plant` = "red", `Scirpus` = "blue", `Triglochin` = "green"),
    `Time` = c(D4 = "purple", D66 = "yellow"),
    Phylum = phylum_col
)

png(here("output", "figures","heatmap_D4-D66-on-sediments.png"),width=8,height=4,units="in",res=600)
ComplexHeatmap::pheatmap(matrix, scale = "row", 
                         fontface_row = "italic",
                         show_colnames = FALSE,
                         fontsize = 8, 
                         fontsize_row = 8,       # Font size for row names
                         fontsize_col = 8, 
                         annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = ann_colors)
dev.off()
```


## Volcano genus

https://bioconductor_org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano_html#introduction
```{r}
  # if (!requireNamespace('BiocManager', quietly = TRUE))
  #   install_packages('BiocManager')
  # 
  # BiocManager::install('EnhancedVolcano')

res_df <- as.data.frame(res)
taxa_sig_df <- tax_clean[rownames(tax_clean) %in% rownames(res_df),]
#library(EnhancedVolcano)
res_volcano <- merge(res_df, taxa_sig_df, by = "row.names")

# Extract differentially abundant genera: 
# cutoffs for log2FC: abs(log2FC) > 1 and p-adjust: -log10(padj) > 5) for padj = 0.00001

# Convert p-values to -log10 scale
res_volcano$neg_log10_pvalue <- -log10(res_volcano$padj)

# Filter res_volcano based on cutoffs
filtered_res_volcano <- res_volcano[(abs(res_volcano$log2FoldChange) > 1) & (res_volcano$neg_log10_pvalue > 5), ]

# Create a new dataframe with the relevant columns
result_genus_diff <- filtered_res_volcano[, c("Genus", "log2FoldChange", "neg_log10_pvalue")]
result_genus_diff$Genus
# View the resulting dataframe
print(result_df)




volcano_sed_4.66_genus <- EnhancedVolcano(res_volcano, 
                          lab = res_volcano$Genus, 
                          x = 'log2FoldChange', 
                          y = "padj",
                          title = "",
                          subtitle = "D4 vs D66 in Sediments - Genus",
                          pointSize = 4.0,
                          labSize = 3.0,
                          colAlpha = 0.5,
                          legendPosition = 'top',
                          legendLabSize = 10,
                          legendIconSize = 4.0,
                          drawConnectors = TRUE,
                          widthConnectors = 0.6,
                          max.overlaps = 20
                          )

volcano_sed_4.66_genus

ggsave(filename = here("output", "figures", "deseq_volcano_sed_4.66_genus.png"), plot = volcano_sed_4.66_genus, width = 6, height = 6, dpi = 300)
ggsave(filename = here("output", "figures", "deseq_volcano_sed_4.66_genus.pdf"), plot = volcano_sed_4.66_genus, width = 6, height = 6, dpi = 300)


```





# -------------------------------------------------------

# DESeq2 -  ORDER

# -------------------------------------------------------



## Agglom at Order level
```{r}
sample_data(ps)$Time <- as.factor(sample_data(ps)$Time) # factorize for DESeq2
ps_taxa <- tax_glom(ps, taxrank = 'Order', NArm = FALSE) # 323
```

```{r}
# Subset the original phyloseq object 'ps_taxa' to only include samples from time points "D4" and "D66"
ps_taxa_sub <- subset_samples(ps_taxa, Time %in% c("D4", "D66"))

# Filter out features (OTUs or taxa) that have more than 90% zeros across all samples
# This is done by checking the number of zeros in each row of the OTU table and comparing it to 90% of the total number of columns (samples)
ps_taxa_pse_sub <- prune_taxa(rowSums(otu_table(ps_taxa_sub) == 0) < ncol(otu_table(ps_taxa_sub)) * 0.9, ps_taxa_sub) #223

# Convert the filtered phyloseq object to a DESeq2 object, using 'Time' as the variable of interest
ps_ds <- phyloseq_to_deseq2(ps_taxa_pse_sub, ~ Time)

# Estimate size factors for normalization using the "poscounts" method, which is useful when many genes have zeros in some samples
ds <- estimateSizeFactors(ps_ds, type = "poscounts")

# Perform differential abundance analysis using the Wald test and a parametric fit
ds <- DESeq(ds, test = "Wald", fitType = "parametric")

# Set the significance threshold (alpha) to 0_05
alpha <- 0.05 

# Retrieve the results of the differential abundance analysis
res <- results(ds, alpha = alpha)

# Order the results by the adjusted p-values (padj) in ascending order
res <- res[order(res$padj, na.last = NA), ]
res_df1 <- res
# Extract the names (taxa) of the top 20 features with the lowest adjusted p-values
taxa_sig <- rownames(res[1:20, ])

# Convert the counts in the original phyloseq object 'ps_taxa' to relative abundances (percentages)
ps_taxa_rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)

# Subset the relative abundance phyloseq object to only include the significant taxa identified earlier
ps_taxa_rel_sig <- prune_taxa(taxa_sig, ps_taxa_rel)

# Further subset the relative abundance phyloseq object to only include samples from time points "D4" and "D66"
ps_taxa_rel_sig <- prune_samples(colnames(otu_table(ps_taxa_pse_sub)), ps_taxa_rel_sig)

```


## Volcano plot

https://bioconductor_org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano_html#introduction
```{r}
  # if (!requireNamespace('BiocManager', quietly = TRUE))
  #   install_packages('BiocManager')
  # 
  # BiocManager::install('EnhancedVolcano')

res_df <- as.data.frame(res)
taxa_sig_df <- tax_clean[rownames(tax_clean) %in% rownames(res_df),]
library(EnhancedVolcano)
res_volcano <- merge(res_df, taxa_sig_df, by = "row.names")

volcano_sed_4.66_order <- EnhancedVolcano(res_volcano, 
                          lab = res_volcano$Order, 
                          x = 'log2FoldChange', 
                          y = "padj",
                          title = "",
                          subtitle = "D4 vs D66 in Sediments",
                          pointSize = 4.0,
                          labSize = 3.0,
                          colAlpha = 0.5,
                          legendPosition = 'top',
                          legendLabSize = 10,
                          legendIconSize = 4.0,
                          drawConnectors = TRUE,
                          widthConnectors = 0.6,)

volcano_sed_4.66_order
ggsave(filename = here("output", "figures", "deseq_volcano_sed_4.66_order.png"), plot = volcano_sed_4.66_order, width = 6, height = 6, dpi = 300)
ggsave(filename = here("output", "figures", "deseq_volcano_sed_4.66_order.pdf"), plot = volcano_sed_4.66_order, width = 6, height = 6, dpi = 300)

```






